### ISSUE TITLE
andi.w/ori.w to %sr failure

### ISSUE THREAD

**Issue by ggaleotti** [vivier/qemu-m68k#1]
Hi Vivier,
first of all, many thanks for your excellent work in qemu-m68k. I hope that soon your work will become official in the qemu tree.
I'm writing a little "bios" for the m68k, and I was able to start the m68k quadra 800 emulator from
ROM, executing C code and send characters to the serial port. Executing this little fragment (in order to load the msp):
# define MBIT  (1 << 12)

```
            ori.w   #MBIT,%sr                         // select Master SSP
            lea.l   msp_stack,%sp                   // load the stack
            andi.w  #~MBIT,%sr                     // select back the ISP
```

gives me this error:

tcg.c:1981: tcg_reg_alloc_op: Assertion `ts->val_type == 1' failed.

I've tried some changes, but no matter what instructions you put between the ori/andi, the result is the same. Unfortunately I am not confident with the innards of the qemu tcg, so I am kindly filing this issue.
Many thanks again and best regards.
G

**Comment by vivier** [comment]
Could you run qemu with "-d in_asm,op" to have the TCG traces ?

**Comment by ggaleotti** [comment]
sorry, forgot about a trace ...

qemu-system-m68k  -M q800 -m 8 -L /usr/local/share/qemu -bios MacROM.bin -nographic -d in_asm,op -serial /dev/ttyVS0
QEMU 2.3.0 monitor - type 'help' for more information
(qemu) ----------------
IN: 
0x40000400:  movew #9984,%sr

OP:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x40000400
 movi_i32 tmp0,$0x2700
 movi_i32 tmp2,$0xf
 and_i32 CC_DEST,tmp0,tmp2
 movi_i32 tmp2,$0x4
 shr_i32 tmp1,tmp0,tmp2
 movi_i32 tmp2,$0x1
 and_i32 CC_X,tmp1,tmp2
 call set_sr,$0x0,$0,env,tmp0
 discard CC_SRC
 discard CC_OP
 movi_i32 CC_OP,$0x1
 movi_i32 PC,$0x40000404
 movi_i32 CC_OP,$0x1
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5c1500b

---

IN: 
0x40000404:  clrl %d0
0x40000406:  movec %d0,%sfc

OP:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x40000404
 movi_i32 tmp0,$0x0
 mov_i32 D0,tmp0
 mov_i32 CC_DEST,tmp0
 discard CC_SRC
 discard CC_OP

 ---- 0x40000406
 movi_i32 tmp1,$0x0
 call movec_to,$0x0,$0,env,tmp1,D0
 movi_i32 CC_OP,$0x4
 movi_i32 PC,$0x4000040a
 movi_i32 CC_OP,$0x4
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5c1504f

---

IN: 
0x4000040a:  movec %d0,%dfc

OP:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x4000040a
 movi_i32 tmp0,$0x1
 call movec_to,$0x0,$0,env,tmp0,D0
 movi_i32 PC,$0x4000040e
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5c15093

---

IN: 
0x4000040e:  movec %d0,%cacr

OP:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x4000040e
 movi_i32 tmp0,$0x2
 call movec_to,$0x0,$0,env,tmp0,D0
 movi_i32 PC,$0x40000412
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5c150d7

---

IN: 
0x40000412:  movec %d0,%tc

OP:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x40000412
 movi_i32 tmp0,$0x3
 call movec_to,$0x0,$0,env,tmp0,D0
 movi_i32 PC,$0x40000416
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5c1511b

---

IN: 
0x40000416:  movec %d0,%itt0

OP:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x40000416
 movi_i32 tmp0,$0x4
 call movec_to,$0x0,$0,env,tmp0,D0
 movi_i32 PC,$0x4000041a
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5c1515f

---

IN: 
0x4000041a:  movec %d0,%itt1

OP:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x4000041a
 movi_i32 tmp0,$0x5
 call movec_to,$0x0,$0,env,tmp0,D0
 movi_i32 PC,$0x4000041e
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5c151a3

---

IN: 
0x4000041e:  movec %d0,%dtt0

OP:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x4000041e
 movi_i32 tmp0,$0x6
 call movec_to,$0x0,$0,env,tmp0,D0
 movi_i32 PC,$0x40000422
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5c151e7

---

IN: 
0x40000422:  movec %d0,%dtt1

OP:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x40000422
 movi_i32 tmp0,$0x7
 call movec_to,$0x0,$0,env,tmp0,D0
 movi_i32 PC,$0x40000426
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5c1522b

---

IN: 
0x40000426:  lea 0x800,%sp
0x4000042c:  oriw #4096,%sr
0x40000430:  lea 0x900,%sp
0x40000436:  andiw #-4097,%sr
0x4000043a:  jsr 0x400006b6

OP:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x40000426
 movi_i32 tmp0,$0x800
 mov_i32 A7,tmp0

 ---- 0x4000042c
 movi_i32 tmp1,$0x1000
 call flush_flags,$0x0,$1,tmp2,env,CC_OP
 movi_i32 tmp4,$0x4
 shl_i32 tmp3,CC_X,tmp4
 or_i32 tmp3,tmp3,tmp2
 movi_i32 tmp5,$0xffe0
 and_i32 tmp4,SR,tmp5
 or_i32 tmp4,tmp4,tmp3
 or_i32 tmp5,tmp4,tmp1
 movi_i32 tmp8,$0xf
 and_i32 CC_DEST,tmp5,tmp8
 movi_i32 tmp8,$0x4
 shr_i32 tmp7,tmp5,tmp8
 movi_i32 tmp8,$0x1
 and_i32 CC_X,tmp7,tmp8
 call set_sr,$0x0,$0,env,tmp5
 discard CC_SRC
 discard CC_OP

 ---- 0x40000430
 movi_i32 tmp8,$0x900
 mov_i32 A7,tmp8

 ---- 0x40000436
 movi_i32 tmp9,$0xffffefff
 movi_i32 tmp12,$0x4
 shl_i32 tmp11,CC_X,tmp12
 or_i32 tmp11,tmp11,tmp10
 movi_i32 tmp13,$0xffe0
 and_i32 tmp12,SR,tmp13
 or_i32 tmp12,tmp12,tmp11
 and_i32 tmp13,tmp12,tmp9
 movi_i32 tmp16,$0xf
 and_i32 CC_DEST,tmp13,tmp16
 movi_i32 tmp16,$0x4
 shr_i32 tmp15,tmp13,tmp16
 movi_i32 tmp16,$0x1
 and_i32 CC_X,tmp15,tmp16
 call set_sr,$0x0,$0,env,tmp13

 ---- 0x4000043a
 movi_i32 tmp16,$0x400006b6
 movi_i32 tmp17,$0x40000440
 movi_i32 tmp19,$0x4
 sub_i32 tmp18,A7,tmp19
 movi_i32 CC_OP,$0x1
 qemu_st_i32 tmp17,tmp18,beul,$0x0
 mov_i32 A7,tmp18
 movi_i32 CC_OP,$0x1
 mov_i32 PC,tmp16
 movi_i32 CC_OP,$0x1
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5c1526f

qemu-system-m68k: /usr/local/src/qemu-m68k-q800-v2.3.0/tcg/tcg.c:1981: tcg_reg_alloc_op: Assertion `ts->val_type == 1' failed.
./startqemu: line 27: 25880 Aborted                 qemu-system-m68k $OPTIONS
serialport_emulator=25878 /dev/ttyVS0=/dev/pts/1 /dev/ttyVS1=/dev/pts/2

**Comment by vivier** [comment]
I think the error is after this piece of code as this is a TCG generation error and the opcodes for this part seems to be correctly generated. What is the code at 0x400006b6 address ?

**Comment by ggaleotti** [comment]
The code at 0x400006b6 is a jsr to a startup routine in C. Some things to note:
1) if I remove the jsr and put a dead loop (xxx: jmp xxx) after the andi.w, I got error again
2) without the offending sequence "ori.w ... andi.w" the rest of the code runs fine
3) if I perform the same operations by means of setting first d0 and then transfer do to sr, everything is ok
4) the hint seems to be exactly the ori.w ... andi.w, if I remove one of the two, the emulator runs fine
Anyway, I will perform some other tests ...

**Comment by vivier** [comment]
OK, you should be right the error seems to be here. Please run with "-d in_asm, op_opt", the issue could be in the optimisation.

**Comment by ggaleotti** [comment]
Ok, here is the dump. Now there is a "dead" jmp after the andi.w so to get rid of other variables in this behaviour. My eyes are shutting (00:46 local). A lot of thanks for your great effort, hope see you very soon.

qemu-system-m68k  -M q800 -m 8 -L /usr/local/share/qemu -bios MacROM.bin -nographic -d in_asm,op_opt -serial /dev/ttyVS0
QEMU 2.3.0 monitor - type 'help' for more information
(qemu) ----------------
IN: 
0x40000400:  movew #9984,%sr

OP after optimization and liveness analysis:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x40000400
 movi_i32 tmp0,$0x2700
 movi_i32 CC_DEST,$0x0
 movi_i32 CC_X,$0x0
 call set_sr,$0x0,$0,env,tmp0
 discard CC_SRC
 discard CC_OP
 movi_i32 PC,$0x40000404
 movi_i32 CC_OP,$0x1
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5bfa00b

---

IN: 
0x40000404:  clrl %d0
0x40000406:  movec %d0,%sfc

OP after optimization and liveness analysis:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x40000404
 movi_i32 D0,$0x0
 movi_i32 CC_DEST,$0x0
 discard CC_SRC
 discard CC_OP

 ---- 0x40000406
 movi_i32 tmp1,$0x0
 call movec_to,$0x0,$0,env,tmp1,D0
 movi_i32 PC,$0x4000040a
 movi_i32 CC_OP,$0x4
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5bfa04f

---

IN: 
0x4000040a:  movec %d0,%dfc

OP after optimization and liveness analysis:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x4000040a
 movi_i32 tmp0,$0x1
 call movec_to,$0x0,$0,env,tmp0,D0
 movi_i32 PC,$0x4000040e
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5bfa093

---

IN: 
0x4000040e:  movec %d0,%cacr

OP after optimization and liveness analysis:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x4000040e
 movi_i32 tmp0,$0x2
 call movec_to,$0x0,$0,env,tmp0,D0
 movi_i32 PC,$0x40000412
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5bfa0d7

---

IN: 
0x40000412:  movec %d0,%tc

OP after optimization and liveness analysis:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x40000412
 movi_i32 tmp0,$0x3
 call movec_to,$0x0,$0,env,tmp0,D0
 movi_i32 PC,$0x40000416
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5bfa11b

---

IN: 
0x40000416:  movec %d0,%itt0

OP after optimization and liveness analysis:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x40000416
 movi_i32 tmp0,$0x4
 call movec_to,$0x0,$0,env,tmp0,D0
 movi_i32 PC,$0x4000041a
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5bfa15f

---

IN: 
0x4000041a:  movec %d0,%itt1

OP after optimization and liveness analysis:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x4000041a
 movi_i32 tmp0,$0x5
 call movec_to,$0x0,$0,env,tmp0,D0
 movi_i32 PC,$0x4000041e
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5bfa1a3

---

IN: 
0x4000041e:  movec %d0,%dtt0

OP after optimization and liveness analysis:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x4000041e
 movi_i32 tmp0,$0x6
 call movec_to,$0x0,$0,env,tmp0,D0
 movi_i32 PC,$0x40000422
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5bfa1e7

---

IN: 
0x40000422:  movec %d0,%dtt1

OP after optimization and liveness analysis:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x40000422
 movi_i32 tmp0,$0x7
 call movec_to,$0x0,$0,env,tmp0,D0
 movi_i32 PC,$0x40000426
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5bfa22b

---

IN: 
0x40000426:  lea 0x800,%sp
0x4000042c:  oriw #4096,%sr
0x40000430:  lea 0x900,%sp
0x40000436:  andiw #-4097,%sr
0x4000043a:  jmp %pc@(0x4000043a)

OP after optimization and liveness analysis:
 ld_i32 tmp0,env,$0xfffffffc
 movi_i32 tmp1,$0x0
 brcond_i32 tmp0,tmp1,ne,$L0

 ---- 0x40000426
 movi_i32 A7,$0x800

 ---- 0x4000042c
 movi_i32 tmp1,$0x1000
 call flush_flags,$0x0,$1,tmp2,env,CC_OP
 movi_i32 tmp4,$0x4
 shl_i32 tmp3,CC_X,tmp4
 or_i32 tmp3,tmp3,tmp2
 movi_i32 tmp5,$0xffe0
 and_i32 tmp4,SR,tmp5
 or_i32 tmp4,tmp4,tmp3
 or_i32 tmp5,tmp4,tmp1
 movi_i32 tmp8,$0xf
 and_i32 CC_DEST,tmp5,tmp8
 movi_i32 tmp8,$0x4
 shr_i32 tmp7,tmp5,tmp8
 movi_i32 tmp8,$0x1
 and_i32 CC_X,tmp7,tmp8
 call set_sr,$0x0,$0,env,tmp5
 discard CC_SRC
 discard CC_OP

 ---- 0x40000430
 movi_i32 A7,$0x900

 ---- 0x40000436
 movi_i32 tmp9,$0xffffefff
 movi_i32 tmp12,$0x4
 shl_i32 tmp11,CC_X,tmp12
 or_i32 tmp11,tmp11,tmp10
 movi_i32 tmp13,$0xffe0
 and_i32 tmp12,SR,tmp13
 or_i32 tmp12,tmp12,tmp11
 and_i32 tmp13,tmp12,tmp9
 movi_i32 tmp16,$0xf
 and_i32 CC_DEST,tmp13,tmp16
 movi_i32 tmp16,$0x4
 shr_i32 tmp15,tmp13,tmp16
 movi_i32 tmp16,$0x1
 and_i32 CC_X,tmp15,tmp16
 call set_sr,$0x0,$0,env,tmp13

 ---- 0x4000043a
 movi_i32 PC,$0x4000043a
 movi_i32 CC_OP,$0x1
 exit_tb $0x0
 set_label $L0
 exit_tb $0xb5bfa26f

qemu-system-m68k: /usr/local/src/qemu-m68k-q800-v2.3.0/tcg/tcg.c:1981: tcg_reg_alloc_op: Assertion `ts->val_type == 1' failed.
./startqemu: line 28: 26619 Aborted                 qemu-system-m68k $OPTIONS
serialport_emulator=26617 /dev/ttyVS0=/dev/pts/1 /dev/ttyVS1=/dev/pts/2

**Comment by vivier** [comment]
Could you provide me the binary of your ROM ? I need to do more tests to understand what happens.

**Comment by ggaleotti** [comment]
Absolutely yes, of course. In the attachment you can find, other than the ROM, a gzipped mini-project with the sources. It is cutted down to the bare minimum, but it will suffice to generate the failing binary. I use an m68k-unknown-elf cross toolchain (binutils 2.25, gcc 5.1.0), I hope you have something similar. Thank you Laurent.Gabriele

```
 Il Giovedì 6 Agosto 2015 1:34, Laurent Vivier <notifications@github.com> ha scritto:
```

 Could you provide me the binary of your ROM ? I need to do more tests to understand what happens.—

**Comment by vivier** [comment]
Nice bug, thank you !

Fixed by commit:

2bf3d3a m68k: copy the CC_DEST register to the TCGv register

Available on branch q800-v2.3.0

**Comment by ggaleotti** [comment]
Thanks a lot, very nice work!
